crankshaft-docker/src/container.rs

```diff
@@ -43,6 +51,75 @@ pub use builder::Builder;
 /// allocations.
 const DEFAULT_TAR_CAPACITY: usize = 0xFFFF;
 
+/// The default retry strategy for fallable Docker operations.
+///
+/// Docker operations are often flaky due to the state of the system or the
+/// Docker daemon. This retry duration is used as a baseline for all operations
+/// that require communication with the Docker daemon.
+///
+/// Use the [`default_retry`] function to easily use this retry strategy.
+///
+/// See https://github.com/stjude-rust-labs/crankshaft/issues/68 for more
+/// information.
+fn default_retry_strategy() -> impl Iterator<Item = Duration> {
+    ExponentialBackoff::from_millis(50)
+        .factor(2)
+        .max_delay_millis(1000)
+        .map(jitter)
+        .take(5)
+}
+
+/// Helper to determine if a Docker error is retryable or not.
+fn is_retryable(err: bollard::errors::Error) -> RetryError<bollard::errors::Error> {
+    match err {
+        // A transient I/O issue.
+        bollard::errors::Error::IOError { .. } |
+        // A transient connection issue.
+        bollard::errors::Error::HyperResponseError { .. } => RetryError::transient(err),
+        // A docker server error, where...
+        bollard::errors::Error::DockerResponseServerError { status_code, .. }
+          // Docker reports a conflict, which includes things like trying to
+          // stop a container that the daemon doesn't _think_ is currently
+          // stopped.
+          if status_code == 409 => {
+            RetryError::transient(err)
+          }
+        _ => {
+            RetryError::permanent(err)
+        }
+    }
+}
+
+/// Helper to perform the Docker operation with the default retry strategy.
+async fn default_retry<F, Fut, T>(op: F) -> std::result::Result<T, bollard::errors::Error>
+where
+    F: Fn() -> Fut,
+    Fut: Future<Output = std::result::Result<T, bollard::errors::Error>>,
+{
+    let action = || async { op().await.map_err(is_retryable) };
+    Retry::spawn(default_retry_strategy(), action).await
+}
+
+/// The default timeout for an operation.
+///
+/// Use the [`default_timeout`] function to easily use this duration.
+///
+/// This value was picked because it's a reasonable compromise of being a
+/// sufficiently long amount of time while not being too long for operations
+/// occuring with containers.
+const DEFAULT_TIMEOUT: Duration = Duration::from_secs(3);
+
+/// Helper to perform the operation with a default timeout.
+#[expect(unused)]
```

Comment:
Just noting that I went to the trouble of creating methods for timeouts, but I ended up not needing them (see the big block of commented code below). As such, I just kept them in here to keep us from having to write them again.